nemo/collections/nlp/models/language_modeling/megatron/gpt_full_te_layer_autocast_spec.py

```diff
@@ -0,0 +1,245 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Optional
+
+import torch
+from megatron.core import parallel_state, tensor_parallel
+from megatron.core.transformer.spec_utils import ModuleSpec
+from megatron.core.transformer.transformer_block import TransformerBlockSubmodules
+from megatron.core.transformer.utils import make_sharded_tensors_for_checkpoint
+
+from transformer_engine.pytorch import TransformerLayer
+
+from nemo.collections.nlp.parts import utils_funcs
+
+
+# Copied from nemo/collections/nlp/modules/common/megatron/transformer.py
+# as the source file is slated to be removed
+class AutocastTransformerLayer(TransformerLayer):
+    def __init__(
+        self,
+        hidden_size: int,
+        ffn_hidden_size: int,
+        layernorm_epsilon: float,
+        num_attention_heads: int,
+        init_method: Callable,
+        output_layer_init_method: Callable,
+        hidden_dropout: float,
+        attention_dropout: float,
+        layer_number: Optional[int] = None,
+        kv_channels: Optional[int] = None,
+        self_attn_mask_type: str = "causal",
+        tp_group: Optional[Any] = None,
+        tp_size: int = 1,
+        params_dtype: torch.dtype = torch.float32,
+        get_rng_state_tracker: Optional[Callable] = None,
+        fuse_wgrad_accumulation: bool = False,
+        seq_length: Optional[int] = None,
+        micro_batch_size: Optional[int] = None,
+        sequence_parallel: bool = False,
+        apply_residual_connection_post_layernorm: bool = False,
+        output_layernorm: bool = False,
+        layer_type: str = "encoder",
+        drop_path_rate: float = 0,
+        use_emha: bool = False,
+        ub_tp_comm_overlap: bool = False,
```

Comment:
Still needs the TP knob changes from [here](https://github.com/NVIDIA/NeMo/pull/8290/files), in case you missed it

Reply:
@jbaczek I think you misunderstood my comment, the TP knobs have to be updated in our copy of `AutocastTransformerLayer` (in `nemo/collections/nlp/models/language_modeling/megatron/gpt_full_te_layer_autocast_spec.py`), the [NeMo PR](https://github.com/NVIDIA/NeMo/pull/8290) that I mentioned is now merged so we don't need those changes.

Reply:
Got it! I've added these knobs. I leave modifications to `transformer.py`. It will be easier to test and merge it in the future. @minitu can you review this?